src/abstractdataframe/show.jl

```diff
@@ -573,61 +573,213 @@ function _show(io::IO,
+    indices = Vector{Int}[]
+    padding = Vector{Int}[]
+
+    # If the screen is limited, we do not need to process all the numbers.
+    dsize = displaysize(io)
+
+    if !allcols
+        # Given the spacing, there is no way to fit more than W/9 rows of
+        # floating numbers in the screen, where W is the display width.
+        Δc = clamp(div(dsize[2], 9), 0, num_cols)
+    else
+        Δc = num_cols
+    end
+
+    if !allrows
+        # Get the maximum number of lines that we can display given the screen size.
+        Δr = clamp(dsize[1] - 4, 0, num_rows)
+    else
+        Δr = num_rows
+    end
+
+    Δr_lim = cld(Δr, 2)
+
+    # Columns composed of numbers are printed aligned to the left.
+    for i = 1:Δc
+        type_i = nonmissingtype(types[i])
+
+        if type_i <: AbstractFloat
+            alignment[i] = :r
+            push!(float_cols, i)
+        elseif type_i <: Number
+            alignment[i] = :r
         end
     end
-    dsize = displaysize(io)
-    availableheight = dsize[1] - 7
-    nrowssubset = fld(availableheight, 2)
-    bound = min(nrowssubset - 1, nrows)
-    if allrows || nrows <= availableheight
-        rowindices1 = 1:nrows
-        rowindices2 = 1:0
+
+    # Check if the quantity of data to be printed allows the alignment of
+    # floats.
+    num_float_cols = length(float_cols)
+
+    if Δr*num_float_cols ≤ 200_000
+        for i in float_cols
+            # Analyze the order of the number to compute the maximum padding
+            # that must be applied to align the numbers at the decimal
+            # point.
+
+            max_pad_i = 0
+            order_i = zeros(Δr)
+            indices_i = zeros(Δr)
+
+            col = df[!, i]
+
+            for k = 1:Δr
+                # We need to process the top and bottom of the table because
+                # we are cropping in the middle.
+
+                kr = k ≤ Δr_lim ? k : num_rows - (k - Δr_lim) + 1
+
+                v = col[kr]
+
+                order_v = 0
+
+                if v isa Number
+                    abs_v = abs(v)
+                    log_v = (v isa Union{Real, Complex} && !isinf(v) && !isnan(v) && abs_v > 1) ?
+                        floor(Int, log10(abs_v))::Int : 0
+
+                    # If the order is higher than 5, then we print using
+                    # scientific notation.
+                    order_v = log_v > 5 ? 0 : floor(Int, log_v)
+
+                    # If the number is negative, we need to add an additional
+                    # padding to print the sign.
+                    v < 0 && (order_v += 1)
+                end
+
+                order_i[k] = order_v
+                indices_i[k] = kr
+
+                order_v > max_pad_i && (max_pad_i = order_v)
+            end
+
+            # The algorithm requires the cells to be left aligned.
+            alignment[i] = :l
+
+            push!(indices, indices_i)
+            push!(padding, max_pad_i .- order_i)
```

Comment:
You can avoid a copy by doing `order_i .= max_pad_i .- order_i` and then passing `order_i`.

Reply:
Ok!